openflow13/nx_action.go

```diff
@@ -0,0 +1,953 @@
+	NXAST_POP_QUEUE        = 5
+	NXAST_REG_MOVE         = 6
+	NXAST_REG_LOAD         = 7
+	NXAST_NOTE             = 8
+	NXAST_SET_TUNNEL_V6    = 9
+	NXAST_MULTIPATH        = 10
+	NXAST_AUTOPATH         = 11
+	NXAST_BUNDLE           = 12
+	NXAST_BUNDLE_LOAD      = 13
+	NXAST_RESUBMIT_TABLE   = 14
+	NXAST_OUTPUT_REG       = 15
+	NXAST_LEARN            = 16
+	NXAST_EXIT             = 17
+	NXAST_DEC_TTL          = 18
+	NXAST_FIN_TIMEOUT      = 19
+	NXAST_CONTROLLER       = 20
+	NXAST_DEC_TTL_CNT_IDS  = 21
+	NXAST_PUSH_MPLS        = 23
+	NXAST_POP_MPLS         = 24
+	NXAST_SET_MPLS_TTL     = 25
+	NXAST_DEC_MPLS_TTL     = 26
+	NXAST_STACK_PUSH       = 27
+	NXAST_STACK_POP        = 28
+	NXAST_SAMPLE           = 29
+	NXAST_SET_MPLS_LABEL   = 30
+	NXAST_SET_MPLS_TC      = 31
+	NXAST_OUTPUT_REG2      = 32
+	NXAST_REG_LOAD2        = 33
+	NXAST_CNJUNCTION       = 34
+	NXAST_CT               = 35
+	NXAST_NAT              = 36
+	NXAST_CONTROLLER2      = 37
+	NXAST_SAMPLE2          = 38
+	NXAST_OUTPUT_TRUNC     = 39
+	NXAST_CT_CLEAR         = 43
+	NXAST_CT_RESUBMIT      = 44
+	NXAST_RAW_ENCAP        = 46
+	NXAST_RAW_DECAP        = 47
+	NXAST_DEC_NSH_TTL      = 48
+)
+
+type NXActionHeader struct {
+	*ActionHeader
+	Vender  uint32
+	Subtype uint16
+}
+
+func (a *NXActionHeader) Header() *ActionHeader {
+	return a.ActionHeader
+}
+
+func (a *NXActionHeader) NXHeader() *NXActionHeader {
+	return a
+}
+
+func (a *NXActionHeader) Len() (n uint16) {
+	return 10
+}
+
+func (a *NXActionHeader) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, a.Len())
+	b := make([]byte, 0)
+	n := 0
+
+	b, err = a.ActionHeader.MarshalBinary()
+	copy(data[n:], b)
+	n += len(b)
+	binary.BigEndian.PutUint32(data[n:], a.Vender)
+	n += 4
+	binary.BigEndian.PutUint16(data[n:], a.Subtype)
+	return
+}
+
+func (a *NXActionHeader) UnmarshalBinary(data []byte) error {
+	if len(data) != 10 {
+		return errors.New("The byte array has wrong size to unmarshal an NXActionHeader message.")
+	}
+	n := 0
+	err := a.ActionHeader.UnmarshalBinary(data[n:])
+	n += int(a.ActionHeader.Len())
+	a.Vender = binary.BigEndian.Uint32(data[n:])
+	n += 4
+	a.Subtype = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	return err
+}
+
+func NewNxActionHeader(subtype uint16) *NXActionHeader {
+	actionHeader := &ActionHeader{Type: ActionType_Experimenter, Length: NxActionHeaderLength}
+	return &NXActionHeader{ActionHeader: actionHeader, Vender: NxExperimenterID, Subtype: subtype}
+}
+
+// nxast_conjunction
+type NXActionConjunction struct {
+	*NXActionHeader
+	Clause  uint8
+	NClause uint8
+	ID      uint32
+}
+
+// conjunction(ID, Clause/nclause)
+func NewNXActionConjunction(clause uint8, nclause uint8, id uint32) *NXActionConjunction {
+	a := new(NXActionConjunction)
+	a.NXActionHeader = NewNxActionHeader(NXAST_CNJUNCTION)
+	a.Length = a.NXActionHeader.Len() + 6
+	a.Clause = clause
+	a.NClause = nclause
+	a.ID = id
+	return a
+}
+
+func (a *NXActionConjunction) Len() (n uint16) {
+	return 16
```

Comment:
Avoid magic number.

Reply:
All the above comments apply to rest of the code in this file.

Reply:
For every single action/match field, the message length should be a fixed number. If not so, the program might not able to make sure the expected length of a byte array while invoking function MarshalBinanry(). All existent code also use fixed number (or magic number) to represent the length of a match_field/action message.